alpha-commons/src/main/java/at/ac/tuwien/kr/alpha/commons/WeightedAnswerSet.java

```diff
@@ -0,0 +1,154 @@
+		Iterator<Integer> thisDescIterator = this.weightPerLevel.descendingKeySet().iterator();
+		Iterator<Integer> otherDescIterator = other.weightPerLevel.descendingKeySet().iterator();
+		while (true) {
+			// Descend this level down to one with non-zero weight.
+			Integer thisLevel = null;
+			Integer thisWeight = null;
+			while (thisDescIterator.hasNext()) {
+				thisLevel = thisDescIterator.next();
+				thisWeight = this.weightPerLevel.get(thisLevel);
+				if (thisWeight != null && thisWeight != 0) {
+					break;
+				}
+			}
+			// Descend other level down to one with non-zero weight.
+			Integer otherLevel = null;
+			Integer otherWeight = null;
+			while (otherDescIterator.hasNext()) {
+				otherLevel = otherDescIterator.next();
+				otherWeight = other.weightPerLevel.get(otherLevel);
+				if (otherWeight != null && otherWeight != 0) {
+					break;
+				}
+			}
+			if (thisWeight == null && otherWeight == null) {
+				return 0;
+			}
+			if (thisWeight == null) {
+				return otherWeight < 0 ? 1 : -1;
+			}
+			if (otherWeight == null) {
+				return thisWeight < 0 ? -1 : 1;
+			}
+			// Found level and weight for this and other WeightedAnswerSet.
+			if (!thisLevel.equals(otherLevel)) {
+				return thisLevel < otherLevel ? -1 : 1;
+			}
+			if (!thisWeight.equals(otherWeight)) {
+				return thisWeight < otherWeight ? -1 : 1;
+			}
+			// Same level and same weight for both WeightedAnswerSet, now continue with next lower level.
+		}
+	}
+
+	@Override
+	public int compareTo(AnswerSet other) {
+		if (!(other instanceof WeightedAnswerSet)) {
+			return -1;
+		}
+		int basicComparison = super.compareTo(other);
+		if (basicComparison != 0) {
+			return basicComparison;
+		}
+		WeightedAnswerSet otherWeighted = (WeightedAnswerSet) other;
+		return compareWeights(otherWeighted);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(super.hashCode(), weightPerLevel);
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (!(o instanceof WeightedAnswerSet)) {
+			return false;
+		}
+
+		WeightedAnswerSet that = (WeightedAnswerSet) o;
+
+		if (!getPredicates().equals(that.getPredicates())) {
+			return false;
+		}
+
+		if (!getPredicateInstances().equals(that.getPredicateInstances())) {
+			return false;
+		}
+
+		return weightPerLevel.equals(that.weightPerLevel);
+	}
+
+	public String getWeightsAsString() {
+		StringJoiner joiner = new StringJoiner(", ", "[", "]");
+		for (Map.Entry<Integer, Integer> weightPerLevel : weightPerLevel.entrySet()) {
+			joiner.add(weightPerLevel.getValue() + "@" + weightPerLevel.getKey());
+		}
+		return joiner.toString();
+	}
+
+	/**
+	 * Extract weights at levels from string of comma-separated pairs of the form weight@level.
+	 * Multiple weights for the same level are summed-up.
+	 * @param weightAtLevels
+	 * @return a TreeMap containing
+	 */
+	public static TreeMap<Integer, Integer> weightPerLevelFromString(String weightAtLevels) {
```

Comment:
This is the wrong place to parse command-line arguments. The parsing of this string should happen in `CommandLineParser`. Internally, weights per level should only be passed around as some kind of `Map`s.

Reply:
This is a good catch and I was considering different options for where to put this code. The main problem is, that the same functionality is needed in `WeakConstraintsTests` in the tests of the core module. Relying on code from the cli-app module there felt like a bad idea and the other option would be having the code duplicated. What would you suggest for this kind of dependency?